### PR TITLE
Improve support for sentry-java

### PR DESCRIPTION
**Pull request by markushi**
When working on the Java/Android SDK side for spotlight integration https://github.com/getsentry/sentry-java/pull/3166 I stumbled upon two issues which caused spotlight to crash:
* `sentry-java` sends over "Message" events, which contain a top level [message object](https://develop.sentry.dev/sdk/event-payloads/message/), that clashes with the `message: string` of error events
* transactions may contain no spans at all

JS is not my proficiency, happy for any PR feedback!

image

Before opening this PR:
- [x] I added a [Changeset Entry](https://spotlightjs.com/contribute/changesets/) with `pnpm changeset:add`
- [x] I referenced issues that this PR addresses

### CODE REVIEW

**Comment by markushi** [comment]
A few captured samples when spinning our spring boot example:

[application_x_sentry_envelope-1707730398893.txt](https://github.com/getsentry/spotlight/files/14236807/application_x_sentry_envelope-1707730398893.txt)
[application_x_sentry_envelope-1707730421554.txt](https://github.com/getsentry/spotlight/files/14236808/application_x_sentry_envelope-1707730421554.txt)
[application_x_sentry_envelope-1707730422564.txt](https://github.com/getsentry/spotlight/files/14236809/application_x_sentry_envelope-1707730422564.txt)
[application_x_sentry_envelope-1707730424572.txt](https://github.com/getsentry/spotlight/files/14236810/application_x_sentry_envelope-1707730424572.txt)
[application_x_sentry_envelope-1707730433711.txt](https://github.com/getsentry/spotlight/files/14236811/application_x_sentry_envelope-1707730433711.txt)
[application_x_sentry_envelope-1707730434719.txt](https://github.com/getsentry/spotlight/files/14236812/application_x_sentry_envelope-1707730434719.txt)
[application_x_sentry_envelope-1707730435728.txt](https://github.com/getsentry/spotlight/files/14236813/application_x_sentry_envelope-1707730435728.txt)
[application_x_sentry_envelope-1707730436737.txt](https://github.com/getsentry/spotlight/files/14236814/application_x_sentry_envelope-1707730436737.txt)

File: packages/overlay/src/integrations/sentry/components/Events/Event.tsx [hunk 1]
```diff
@@ -1,40 +1,50 @@
-import { SentryErrorEvent, SentryEvent } from '../../types';
+import { SentryEvent } from '../../types';
```
Comment (Shubhdeep12) [inline packages/overlay/src/integrations/sentry/components/Events/Event.tsx:1]:
```suggestion
import { SentryErrorEvent, SentryEvent } from '../../types';
```

File: packages/overlay/src/integrations/sentry/components/Events/Event.tsx [hunk 2]
```diff
@@ -1,40 +1,50 @@
...
+  if (typeof event.message === 'string') {
+    return event.message;
+  } else if (event.message !== undefined && typeof event.message.formatted === 'string') {
+    return event.message.formatted;
+  } else {
+    return '';
+  }
+}
+
+export function EventTitle({ event }: { event: SentryEvent }) {
```
Comment (Shubhdeep12) [inline packages/overlay/src/integrations/sentry/components/Events/Event.tsx:14]:
```suggestion
export function EventTitle({ event }: { event: SentryEvent | SentryErrorEvent }) {
```

File: packages/overlay/src/integrations/sentry/components/Events/Event.tsx [hunk 3]
```diff
@@ -1,40 +1,50 @@
...
   return (
     <>
-      <strong className="font-bold">{event.message}</strong>
+      <strong className="font-bold">{EventMessage(event)}</strong>
     </>
   );
 }
 
-export function EventSummary({ event }: { event: SentryErrorEvent | SentryEvent }) {
+export function EventSummary({ event }: { event: SentryEvent }) {
```
Comment (Shubhdeep12) [inline packages/overlay/src/integrations/sentry/components/Events/Event.tsx:26]:
```suggestion
export function EventSummary({ event }: { event: SentryErrorEvent | SentryEvent }) {
```

File: packages/overlay/src/integrations/sentry/components/Events/Event.tsx [hunk 4]
```diff
@@ -1,40 +1,50 @@
...
       <h3 className="flex flex-col">
-        <strong className="text-xl">{event.message}</strong>
+        <strong className="text-xl">{EventMessage(event)}</strong>
       </h3>
     </div>
   );
 }
 
-export default function Event({ event }: { event: SentryErrorEvent | SentryEvent }) {
+export default function Event({ event }: { event: SentryEvent }) {
```
Comment (Shubhdeep12) [inline packages/overlay/src/integrations/sentry/components/Events/Event.tsx:39]:
```suggestion
export default function Event({ event }: { event: SentryErrorEvent | SentryEvent }) {
```

File: packages/overlay/src/integrations/sentry/types.ts [hunk 5]
```diff
@@ -47,9 +47,9 @@ export type Breadcrumb = {
 
 type CommonEventAttrs = {
   // not always present, but we are forcing it in EventCache
+  type?: 'error' | 'event' | 'message' | 'default';
```
Comment (Shubhdeep12) [inline packages/overlay/src/integrations/sentry/types.ts:50]: hi @markushi  any specific reason for moving this to `commonEventAttrs`?

Comment (Shubhdeep12) [reply packages/overlay/src/integrations/sentry/types.ts:50]: either we should add "transaction" here as well and remove the type from SentryEvent as well.

OR

lets revert this and move these types to SentryErrorEvent again.

**Review by Shubhdeep12 — CHANGES_REQUESTED** [review]
Hi @markushi added some suggestions, let make `type` attribute separate for both types of event - Error event and transaction.

File: packages/overlay/src/integrations/sentry/data/sentryDataCache.ts [hunk 6]
```diff
@@ -124,23 +124,26 @@ class SentryDataCache {
...
+            ...txn.contexts.trace,
+            start_timestamp: txn.start_timestamp,
+            timestamp: txn.timestamp,
+            description: traceCtx.description || txn.transaction,
+            transaction: txn,
+          });
+
+          if (txn.spans) {
+            allSpans.push(
+              txn.spans.map(s => ({
```
Comment (Shubhdeep12) [inline packages/overlay/src/integrations/sentry/data/sentryDataCache.ts:140]: lets spread the array here 
```suggestion
              ...txn.spans.map(s => ({
```

File: packages/overlay/src/integrations/sentry/types.ts [hunk 7]
```diff
@@ -109,7 +116,7 @@ export type Span = {
 
 export type SentryTransactionEvent = CommonEventAttrs & {
   type: 'transaction';
-  spans: Span[];
+  spans?: Span[];
```
[the remark this answers is not in the document]
Comment (markushi) [reply packages/overlay/src/integrations/sentry/types.ts:119]: No unfortunately not, it can be either `null` or empty.

**Review by Shubhdeep12 — APPROVED** [review]
LGTM   🔎

**Pull request merged**